qCC/ccCommandLineCommands.cpp

```diff
@@ -1318,22 +1428,95 @@ bool CommandSubsample::process(ccCommandLineInterface& cmd)
 		
 		for (CLCloudDesc& desc : cmd.clouds())
 		{
+			//calculate octree before subsampling, it is passed to subsampling, so it won't be recalculated there.
+			CCCoreLib::DgmOctree* octree = desc.pc->computeOctree(nullptr, false).data();
+			if (!octree)
+			{
+				return cmd.error("Octree calculation failed, not enough memory?");
+			}
+			CCCoreLib::ReferenceCloud* refCloud = nullptr;
+			ccPointCloud* result = nullptr;
+			unsigned sizeOfInputCloud = desc.pc->size();
+
 			cmd.print(QObject::tr("\tProcessing cloud %1").arg(!desc.pc->getName().isEmpty() ? desc.pc->getName() : "no name"));
-			
-			CCCoreLib::ReferenceCloud* refCloud = CCCoreLib::CloudSamplingTools::subsampleCloudWithOctreeAtLevel(	desc.pc,
-																													static_cast<unsigned char>(octreeLevel),
-																													CCCoreLib::CloudSamplingTools::NEAREST_POINT_TO_CELL_CENTER,
-																													progressDialog.data());
-			if (!refCloud)
+
+			if (byCellSize)
 			{
-				return cmd.error(QObject::tr("Subsampling process failed!"));
+				//calculate OCTREE level for each cloud based on required octree cell size
+				octreeLevel = static_cast<int>(ceil(log(desc.pc->getOwnBB().getMaxBoxDim() / cellSize) / log(2.0)));
 			}
-			cmd.print(QObject::tr("\tResult: %1 points").arg(refCloud->size()));
-			
-			//save output
-			ccPointCloud* result = desc.pc->partialClone(refCloud);
-			delete refCloud;
-			refCloud = nullptr;
+
+			if (byMaxNumberOfPoints)
+			{
+				if (isPercent)
+				{
+					maxNumberOfPoints = static_cast<unsigned>(ceil(sizeOfInputCloud * percent / 100));
+					cmd.print(QObject::tr("\tOutput point target: %1 * %2% = %3").arg(sizeOfInputCloud).arg(percent).arg(maxNumberOfPoints));
+				}
+
+				//calculate OCTREE level for each cloud based on required number of points
+				octreeLevel = CCCoreLib::DgmOctree::MAX_OCTREE_LEVEL;
+				unsigned numberOfPoints = sizeOfInputCloud;
+				//go through max->min untill previous ptsCount and current ptsCount is different then break;
+				do
+				{
+					unsigned currentNumberOfPoints = octree->getCellNumber(octreeLevel);
+					octreeLevel--;
+					if (currentNumberOfPoints != numberOfPoints && numberOfPoints < maxNumberOfPoints)
+					{
+						break;
+					}
+					numberOfPoints = currentNumberOfPoints;
+				}
+				while (numberOfPoints > maxNumberOfPoints && octreeLevel > 0);
+				octreeLevel++;
```

Comment:
```suggestion
				for (int octreeLevel = CCCoreLib::DgmOctree::MAX_OCTREE_LEVEL; octreeLevel > 1; --octreeLevel)
				{
					unsigned currentNumberOfPoints = octree->getCellNumber(octreeLevel);
					if (currentNumberOfPoints != numberOfPoints && currentNumberOfPoints <= maxNumberOfPoints)
					{
						break;
					}
					numberOfPoints = currentNumberOfPoints;
				}
```

Reply:
i cannot use octreeLevel as index inside for loop, because than it will create octreeLevel a local variable.
I've changed it to currentOctreeLevel